crates/red_knot_python_semantic/src/types/unpacker.rs

```diff
@@ -1,126 +1,185 @@
-                            // This index represents the type of the last element that belongs
-                            // to the starred expression, in an exclusive manner.
-                            let starred_end_index = tuple_ty.len(self.db()) - remaining;
-                            // SAFETY: Safe because of the length check above.
-                            let _starred_element_types =
-                                &tuple_ty.elements(self.db())[starred_index..starred_end_index];
-                            // TODO: Combine the types into a list type. If the
-                            // starred_element_types is empty, then it should be `List[Any]`.
-                            // combine_types(starred_element_types);
-                            element_types.push(todo_type!("starred unpacking"));
-
-                            element_types.extend_from_slice(
-                                // SAFETY: Safe because of the length check above.
-                                &tuple_ty.elements(self.db())[starred_end_index..],
-                            );
-                            Cow::Owned(element_types)
-                        } else {
-                            let mut element_types = tuple_ty.elements(self.db()).to_vec();
-                            // Subtract 1 to insert the starred expression type at the correct
-                            // index.
-                            element_types.resize(elts.len() - 1, Type::Unknown);
-                            // TODO: This should be `list[Unknown]`
-                            element_types.insert(starred_index, todo_type!("starred unpacking"));
-                            Cow::Owned(element_types)
+            | ast::Expr::Tuple(ast::ExprTuple { elts, .. }) => {
+                // Initialize the vector of target types, one for each target.
+                //
+                // This is mainly useful for the union type where the target type at index `n` is
+                // going to be a union of types from every union type element at index `n`.
+                //
+                // For example, if the type is `tuple[int, int] | tuple[int, str]` and the target
+                // has two elements `(a, b)`, then
+                // * The type of `a` will be a union of `int` and `int` which are at index 0 in the
+                //   first and second tuple respectively which resolves to an `int`.
+                // * Similarly, the type of `b` will be a union of `int` and `str` which are at
+                //   index 1 in the first and second tuple respectively which will be `int | str`.
+                let mut target_types = vec![vec![]; elts.len()];
+
+                let unpack_types = match value_ty {
+                    Type::Union(union_ty) => {
+                        Either::Left(union_ty.elements(self.db()).iter().copied())
+                    }
+                    _ => Either::Right(std::iter::once(value_ty)),
+                };
```

Comment:
This is the main entry point that handles the union type. For a union type, we need to apply the logic for each subtype in it.

I want to spend some time to avoid the nested vectors but this is not a hot path so I'd be fine moving with this as well.

Reply:
I actually started with having a vector of `UnionBuilder` but the builder takes ownership of itself when adding the type which isn't possible without multiple pop and push to get the ownership of the builder.